README.md
# mongoclikker

A clean [MongoDB](http://mongodb.org) viewer and basic editor written in [Node](http://nodejs.org). Just clone or download mongoclikker, run `npm install` and fire up [mongoclikker](https://github.com/semu/mongoclikker) with 

    [sudo] node app.js

Open [http://HOSTNAME:2002/view/DATABASE](http://HOSTNAME:2002/view/DATABASE) in your web browser and you are good to go…

![mongoclikker](http://img.hazelco.de/mongoclikker.jpg)

## Security

Per default mongoclikker is protected with a basic HTTP authentication, you should change the default password set in `app.js`:

    app.js:4 protectWith('mongo', 'clikker')

## Edit Mode

[mongoclikker](https://github.com/semu/mongoclikker) supports simple inline edit for basic document properties. 

## Configure

See `app.js` for setting a default database and mongodb connection. Default settings should work fine for getting connected to your local mongodb server. [mongoclikker](https://github.com/semu/mongoclikker) is protected with a simple HTTP access control, read and write access cannot be done without propper authentication.

    require('./mongoclikker.js')/*.setHost('localhost')
            .andPort(27017)
            .forDatabase('mongoclikker') */
            .protectWith('mongo', 'clikker') /* Password for HTTP Authorization */
            .listenOn(2002);
    
## Database Design

[mongoclikker](https://github.com/semu/mongoclikker) works best with documents having an `_id` and a `name` property, included sub documents in arrays or properties should be designed the same way. Of course [mongoclikker](https://github.com/semu/mongoclikker) works fine for collections and databases with different designs!


CHANGELOG.md
## 0.2.2 (current beta)

* new feature: included lesscss.org framework
* enhancement: moved static files in seperate folder
* new tests: data casting
* new feature: sub document display
* enhancement: array display
* new feature: inline edit for basic properties (tested with Google Chrome)
* upstream: includes [connect-basic-auth](https://github.com/semu/connect-basic-auth) with custom bugfix

## 0.2.1 (2011-07-27)

* bugfix: handling for native mongodb driver

## 0.2.0 (2011-07-25)

* new feature: `mongoclikker` displays mongodb data


test/cast.js
var mongoclikker = require('../mongoclikker.js');
var assert = require('assert');

exports.testNumbers = function(beforeExit, assert) {
    assert.equal(1.1, mongoclikker.cast('1.1'));
    assert.equal(2, mongoclikker.cast('2'));
};

exports.testStrings = function(beforeExit, assert) {
    assert.equal('String', mongoclikker.cast('String'));
};

lib/semu-connect-basic-auth-c32ee11/test/basicAuth.test.js
var connect   = require("connect"),
    basicAuth = require("basicAuth"),
    base64    = require("base64");

var server = connect.createServer(
  basicAuth(function (user, password) {
    return user === "user" && password == "password";
  }, "baz"),

  function (req, res) {
    setTimeout(function () {
      res.writeHead(200);
      res.end("welcome " + req.headers.remote_user);
    }, 100);
  }
);

function encode(user, password) {
  return "Basic " + base64.encode(user + ":" + password);
}

module.exports = {
  "test valid credentials": function (assert) {
    var headers = {authorization: encode("user", "password")};

    assert.response(server,
      {url: "/", headers: headers},
      {status: 200, body: "welcome user"}
    );
  },

  "test invalid credentials": function (assert) {
    var headers = {authorization: encode("foo", "bar")};

    assert.response(server,
      {url: "/", headers: headers},
      {status: 401, headers: {"WWW-Authenticate": 'Basic realm="baz"'}}
    );
  },

  "test missing Authorization": function (assert) {
    assert.response(server,
      {url: "/"},
      {status: 401, headers: {"WWW-Authenticate": 'Basic realm="baz"'}}
    );
  },

  "test not a basic request": function (assert) {
    var headers = {authorization: "Foo bar"};

    assert.response(server,
      {url: "/", headers: headers},
      {status: 400}
    );
  }
}


static/script.js
function triggerCancelEvent(curThis, value) { 
  $(curThis).html(value);
  $(curThis).removeClass('isActive');
  $(curThis).click(function(e2) { hasClicked(this, e2); });
}

function triggerSaveEvent(curThis, newValue) {
  $(curThis).html($(curThis).find('input')).append($('<img src="/load.gif" class="loader" />'));
  $.post("/update/" + $(curThis).attr('id'), { data: newValue},
    function(data) {
      $(curThis).removeClass('isActive');
      $(curThis).html(data); } );
  $(curThis).click(function(e2) { hasClicked(this, e2); });
}

function hasClicked(newThis) {
  var curThis = newThis;
  if (!$(curThis).hasClass('isActive')) {
    $(curThis).unbind();
    var tmpText = $(curThis).text();
    $(curThis).html($('<input type="text" value="" class="curInput" /> <span class="saveOptions">[<a href="#" class="save">save</a>] [<a href="#" class="cancel" oldvalue="' + $(curThis).text() + '">cancel</a>]</span>'));
    $(curThis).find('input').val(tmpText);
    $(curThis).find('input').keydown(function(event) {
      if (event.keyCode == '13') { /* save    */ triggerSaveEvent(curThis, $(curThis).find('input').val()); }
      if (event.keyCode == '27') { /* cancel  */ triggerCancelEvent(curThis, $(curThis).find('.cancel').attr('oldvalue')); }
    });
    $(curThis).addClass('isActive');
    $(curThis).find('input').focus();
  }
  
  $(curThis).find('.save').bind('click', (function(e) {
    e.stopImmediatePropagation();
    triggerSaveEvent(curThis, $(curThis).parent().find('input').attr('value'));
  }));
  
  $(curThis).find('.cancel').bind('click', (function(e) {
    e.stopImmediatePropagation();
    triggerCancelEvent(curThis, $(this).attr('oldvalue'));
  }));
}

$(document).ready(function() {
  $('.canEdit').each(function() {
    $(this).click(function(e) {
      e.stopImmediatePropagation();
      hasClicked(this, e); 
    });
  });
});

lib/semu-connect-basic-auth-c32ee11/lib/base64.js
/* adapted by Jed Schmidt for use as a node.js module */

this.encode = base64encode;
this.decode = base64decode;
this.chars = function( string ) {
  base64EncodeChars = string || "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  base64DecodeChars = [];

  for ( var i = 128; i--; ) {
    if ( base64DecodeChars[ i ] === undefined )
      base64DecodeChars[ i ] = -1;

    base64DecodeChars[ base64EncodeChars.charCodeAt( i ) ] = i;
  }

  return this;

};

this.chars();

/* Copyright (C) 1999 Masanao Izumo <iz@onicos.co.jp>
 * Version: 1.0
 * LastModified: Dec 25 1999
 * This library is free.  You can redistribute it and/or modify it.
 */

function base64encode( str ) {
    var out, i, len;
    var c1, c2, c3;

    len = str.length;
    i = 0;
    out = "";
    while(i < len) {
	c1 = str.charCodeAt(i++) & 0xff;
	if(i == len)
	{
	    out += base64EncodeChars.charAt(c1 >> 2);
	    out += base64EncodeChars.charAt((c1 & 0x3) << 4);
	    out += "==";
	    break;
	}
	c2 = str.charCodeAt(i++);
	if(i == len)
	{
	    out += base64EncodeChars.charAt(c1 >> 2);
	    out += base64EncodeChars.charAt(((c1 & 0x3)<< 4) | ((c2 & 0xF0) >> 4));
	    out += base64EncodeChars.charAt((c2 & 0xF) << 2);
	    out += "=";
	    break;
	}
	c3 = str.charCodeAt(i++);
	out += base64EncodeChars.charAt(c1 >> 2);
	out += base64EncodeChars.charAt(((c1 & 0x3)<< 4) | ((c2 & 0xF0) >> 4));
	out += base64EncodeChars.charAt(((c2 & 0xF) << 2) | ((c3 & 0xC0) >>6));
	out += base64EncodeChars.charAt(c3 & 0x3F);
    }
    return out;
}

function base64decode( str ) {
    var c1, c2, c3, c4;
    var i, len, out;

    len = str.length;
    i = 0;
    out = "";
    while(i < len) {
	/* c1 */
	do {
	    c1 = base64DecodeChars[str.charCodeAt(i++) & 0xff];
	} while(i < len && c1 == -1);
	if(c1 == -1)
	    break;

	/* c2 */
	do {
	    c2 = base64DecodeChars[str.charCodeAt(i++) & 0xff];
	} while(i < len && c2 == -1);
	if(c2 == -1)
	    break;

	out += String.fromCharCode((c1 << 2) | ((c2 & 0x30) >> 4));

	/* c3 */
	do {
	    c3 = str.charCodeAt(i++) & 0xff;
	    if(c3 == 61)
		return out;
	    c3 = base64DecodeChars[c3];
	} while(i < len && c3 == -1);
	if(c3 == -1)
	    break;

	out += String.fromCharCode(((c2 & 0XF) << 4) | ((c3 & 0x3C) >> 2));

	/* c4 */
	do {
	    c4 = str.charCodeAt(i++) & 0xff;
	    if(c4 == 61)
		return out;
	    c4 = base64DecodeChars[c4];
	} while(i < len && c4 == -1);
	if(c4 == -1)
	    break;
	out += String.fromCharCode(((c3 & 0x03) << 6) | c4);
    }
    return out;
}


lib/semu-connect-basic-auth-c32ee11/lib/basicAuth.js
var base64 = require("./base64.js");

module.exports = function basicAuth(callback, realm) {
  realm = realm || "Authorization Required";

  function unauthorized(res) {
    res.writeHead(401, {"WWW-Authenticate": 'Basic realm="' + realm + '"'});
    res.end();
  }

  return function(req, res, next) {
    var authorization = req.headers.authorization;

    if (!authorization) {
        return unauthorized(res); }
    
    var parts       = authorization.split(" ");
    var scheme      = parts[0];
    var credentials = base64.decode(parts[1]).split(":");

    if (scheme !== "Basic") {
      res.writeHead(400);
      res.end();
    }

    if (callback(credentials[0], credentials[1]) === true) {
      req.headers["remote_user"] = credentials[0];
      next();
    } else {
      unauthorized(res);
    }
  }
};



mongoclikker.js
/**
 * Default mongoclikker connection settings
 * */
var mongoclikkerConnection = {
  host    : 'localhost'
, port    : '27017'
, user    : ''
, pass    : ''
, db      : 'mongoclikker'
, web     : '2002'
, auth    : {
    user  : ''
  , pass  : ''
  }
};

/**
 * Set hostname for connection
 * @param string host hostname
 * @return object mongoclikker
 * */
var funcSetHost = function(host) {
  mongoclikkerConnection.host = host;
  return this;
};

/**
 * Set port for connection
 * @param integer port port
 * @return object mongoclikker
 * */
var funcAndPort = function(port) {
  mongoclikkerConnection.port = port;
  return this;
};

/**
 * Set username for connection
 * @param string user username
 * @return object mongoclikker
 * */
var funcWithUser = function(user) {
  mongoclikkerConnection.user = user;
  return this;
};

/**
 * Set password for connection
 * @param string password password
 * @return object mongoclikker
 * */
var funcAndPassword = function(password) {
  mongoclikkerConnection.pass = password
  return this;
};

/**
 * Set database name for connection
 * @param string database database name
 * @return object mongoclikker
 * */
var funcforDatabase = function(database) {
  mongoclikkerConnection.db = database;
  return this;
};

/**
 * Set port for web interface
 * @param integer port web interface listens on (Default: 2002)
 * @return object mongoclikker
 * */
var funcAndListenOn = function(port) {
  mongoclikkerConnection.web = port;
  
  funcStartMongoclikker();
};

/**
 * Protect mongoclikker with HTTP Authentication. 
 * @param string user username
 * @param string pass password
 * @return object mongoclikker
 * */
var funcProtectWithUser = function(user, pass) {
  if (!pass || pass == '') {
    throw new Error("Missing password! Mongoclikker has to be started with user and password to avoid external abuse");
  } 
  
  mongoclikkerConnection.auth.user = user;
  mongoclikkerConnection.auth.pass = pass;
  
  return this;
};

/**
 * End mongoclikker interface response
 * @param object res response object
 * */
function endResponse(res) {
  res.write('</html>');
  res.end();
}

/**
 * Try to cast data received from update call
 * @param string data input
 * @return mixed casted output
 * */
function castData(data) {
  var tmpInt = data*1;
  
  if (data == '') {
    return ''; }
  if (data == 'null') {
    return null; }
  if (tmpInt == data) {
    return tmpInt; }
  if (data == 'false') {
    return false; }
  if (data == 'true') {
    return true; }
  return data;
}

/**
 * Get type of given data
 * @param mixed data input
 * @return string type
 * */
function typeOf(data){
  var type = typeof data;
	if (type !== 'object') {
		return type;
	} else if (Object.prototype.toString.call(data) === '[object Array]') {
		return 'array';
	} else if (data === null) {
		return 'null';
	} else {
    if (data instanceof Date) { 
      return 'date'; }
		return 'object';
	}
}

/**
 * Check if given string has mongodb object id format 
 * @param string data input
 * @return boolean
 * */
function isID(data) {
  return (data.length == 12 || data.length == 24);
}

/**
 * Check if edit is supported for given data 
 * @param mixed value data to be displayed
 * @return boolean is editible?
 * */
function canEditValue(value) {
  var curType = typeOf(value);
  return !(curType == 'array' || curType == 'object');
}

/**
 * Display property value in interface 
 * @param mixed data data to be displayed
 * @param string path absolute path for url
 * @param integer index for array values
 * @param object params HTTP params from request
 * @return string styled value ready for output
 * */
function displayValue(data, path, preID, index, params, parentEdit) {
  var curType     = typeOf(data);
  var curValue    = data;
  var curDisplay  = '';
  
  if (curType == 'array') {
    var curLength = curValue.length;
    curDisplay = '<table class="array">';
    for (var i = 0; i < curLength; i++) {
      curDisplay += '<tr><td class="desc key noRightMargin">' + (typeOf(data[i]) == 'array' ? '+' : '-') + ' </td><td class="content value subValue"  id="">' + displayValue(data[i], path, preID, i, params, false) + '</td></tr>';
    }
    curDisplay += '</table>';
  } else if (curType == 'date') {
    curDisplay = curValue;
  } else if (curType == 'object') {
    var hasHeadline = false;
    if (curValue._id && curValue.name) {
      hasHeadline = '<a href="' + path + curValue._id + '">' + curValue.name + ' (#' + curValue._id + ')</a><br />';
    } else if (curValue._id) {
      hasHeadline = '<a href="' + path + curValue._id + '">' + curValue._id + '</a><br />';
    } else if (curValue.name) {
      hasHeadline = '<a href="">' + curValue.name + '</a><br />';
    }

    var curDisplay = hasHeadline || '';
    var objString = curValue.toString();
    if (isID(objString)) {
      curDisplay += '<a href="">' + objString + '</a>';
    } else {
      var styleString = 'display:none;';
      if (params && !hasHeadline || hasHeadline && params.subID && params.subID == curValue._id) {
        styleString = ''; }
      
      curDisplay += '<table style="' + styleString + '" class="array">';
      for (var key in curValue) {
        curDisplay += displayLine(key, curValue[key], path, preID, params, parentEdit);
      }
      curDisplay += '</table>';
    }
  } else {
    curDisplay = curValue;
  }
  
  return curDisplay;
};

/**
 * Display content line with key and value
 * @param sting key item key
 * @param mixed curValue item value
 * @param string path current item path
 * @param string preID current input id path
 * @param mixed params request parameters
 * @return string
 * */
var displayLine = function(key, curValue, path, preID, params, parentEdit) {
  var curID   = preID + '__' + key;
  var canEdit = parentEdit && canEditValue(curValue);
  if (typeOf(curValue) == 'array') {
    key += ' [' +  curValue.length + ']'; }

  return '<tr><td class="desc key">' + key + '</td><td class="content value propValue' + (canEdit ? ' canEdit' : '') + '" id="' + curID + '">' + displayValue(curValue, path + key + '/', curID, 0, params, canEdit) + '</td></tr>';
}

/**
 * Start mongoclikker
 * */
var funcStartMongoclikker = function() {
  var app = require('express').createServer()
    , express = require('express')
    , Db = require('mongodb').Db
    , Connection = require('mongodb').Connection
    , Server = require('mongodb').Server
    , BSON = require('mongodb').BSONNative
    , connectionSettings = {native_parser:true}
    , viewURL = '/view/'
    , ConnectAuth = require(__dirname + '/lib/semu-connect-basic-auth-c32ee11/lib/basicAuth');
    
  var currentDatabase = mongoclikkerConnection.db
    , currentHostname = mongoclikkerConnection.host
    , currentPort     = mongoclikkerConnection.port;
    
  app.use(ConnectAuth(function (user, password) { return user === mongoclikkerConnection.auth.user && password == mongoclikkerConnection.auth.pass; })); 
  app.use(express.bodyParser());  
  app.use(express.static(__dirname + '/static'));
  
  /**
   * Fallback for non-native BSON objects
   * */
  if (BSON == null) {
    BSON = require('mongodb').BSONPure;
    connectionSettings = {};
  }
  
  /**
   * Redirect / to viewURL
   * */
  app.get('/', function(req, res) {
    res.redirect(viewURL);
  });
  
  /**
   * Handle ajax update calls
   * */
  app.post('/update/*', function(req, res) {
    res.writeHead(200, { 'Content-Type': 'text/html' });
    var newData = req.body.data;
    var tmpProp = req.url.replace('/update/', '').split('__');
    
    var objectFind = {'_id': tmpProp[2]};
    if (tmpProp[3].length == 12 || tmpProp[2].length == 24) { 
      objectFind = {'_id': new  BSON.ObjectID(tmpProp[2])}; }
    
    db = new Db(tmpProp[0], new Server(currentHostname, currentPort, {}), connectionSettings)
    db.open(function(err, ignored) {
      if (err) { throw new Error(err); }
      db.collection(tmpProp[1], function(err, collection) {
        var updateKey = tmpProp[3];
        var updateSet = {};
        var dataType  = tmpProp.pop();
        
        updateSet[updateKey] = castData(newData, dataType);
        
        res.write(updateSet[updateKey] + "");
        res.end();
        
        if (tmpProp[4] && tmpProp[5]) {
          /**
           * Update sub document 
           * */
          /*
          var subKey = tmpProp[3] + '._id';
          objectFind[subKey] = tmpProp[4];
          if (tmpProp[4].length == 12 || tmpProp[4].length == 24) { 
            objectFind[subKey] = new BSON.ObjectID(tmpProp[4]); }
          var updateSubKey = tmpProp[3] + '.$.' + tmpProp[5];
          var updateSet = {};
          updateSet[updateSubKey] = newData;
          */
        } 
        
        collection.update(objectFind, {$set: updateSet}, {safe:true},
          function(err) {
            if (err) console.warn(err.message);
          }
        );
      });      
    });
  });
  
  /**
   * Handle database access
   * */
  app.get(viewURL + ':curDB?/:curCollection?/:curStart?/:curLimit?/:curDocument?/:curKey?/:subID?', function(req, res, next) {
    res.write('<html><head><title>mongoclikker</title><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/><script src="https://ajax.googleapis.com/ajax/libs/jquery/1.6.4/jquery.min.js"></script><script src="/script.js"></script><style>@import url("http://fonts.googleapis.com/css?family=Varela+Round&v2");</style><link rel="stylesheet/less" type="text/css" href="/styles.less"><script src="/less.js" type="text/javascript"></script></head>');
  
    /**
     * don't know how to get `show dbs` with node yet, so start with one datebase
     * */
    var dbName = req.params.curDB || currentDatabase
      , listDB = [dbName]
      , db = new Db(dbName, new Server(currentHostname, currentPort, {}), connectionSettings)
      , path = viewURL;
    
    res.write('<table><tr><td class="desc">database</td><td class="content"><ul id="db">');
    for (var i = 0; i < listDB.length; i++) { res.write('<li><a href="' + path + listDB[i] + '">' + listDB + '</a></li>'); }
    res.write('</ul></td></tr>');

    if (!req.params.curDB) {
      /**
       * No database selected
       * */
      endResponse(res);
    } else {
      /**
       * Selected database
       * */
      path += req.params.curDB + '/';
      db.open(function(err, ignored) {
        if (err) { console.log(err); }
        db.collectionNames(function(err, names) {
          var colItems = [];
          res.write('<tr><td class="desc">collection</td><td class="content"><ul id="collection">');
          for (var i = 0; i < names.length; i++) {
            var tmp = names[i].name.replace(req.params.curDB + '.', '');
            if (tmp.indexOf('.indexes') == -1) { 
              res.write('<li><a href="' + path + tmp + '/0/20">' + tmp + '</a></li>'); }
          }
          
          if (!req.params.curCollection) {
            /**
             * No collection selected
             * */
            endResponse(res);
          } else {
            /**
             * Selected collection
             * */
            path += req.params.curCollection + '/' + req.params.curStart + '/' + req.params.curLimit + '/';
            db.collection(req.params.curCollection, function(err, collection) {
              collection.find({}, {'skip':req.params.curStart, 'limit':req.params.curLimit}).toArray(function(err, results) {
                res.write('<tr><td class="desc">item</td><td class="content"><ul id="documents">');
                for (var i = 0; i < results.length; i++) { 
                  if (results[i]) { res.write('<li><a href="' + path + results[i]._id + '">' + (results[i].name && typeof(results[i].name) == 'string' ? results[i].name + ' (#' + results[i]._id + ')' : results[i]._id) + '</a></li>'); } }
                res.write('</ul></td></tr>');
                var prevStart = (req.params.curStart*1 - 1*req.params.curLimit);
                if (prevStart < 0) { 
                  prevStart = 0;  }
                if (req.params.curDocument) { 
                  selectedItem = '/' + req.params.curDocument; }
                
                /**
                 * Build navigation items
                 * */
                var selectedItem = ''
                  , baseURL = viewURL + req.params.curDB + '/' + req.params.curCollection + '/'
                  , nextURL = baseURL + (req.params.curStart*1 + 1*req.params.curLimit) + '/' + req.params.curLimit + selectedItem
                  , prevURL = baseURL + prevStart + '/' + req.params.curLimit + selectedItem;
                res.write('<tr class="docsNav"><td class="desc"></td><td class="content">');
                if (req.params.curStart > 0) { 
                  res.write('<a href="' + prevURL + '">prev</a> | '); }
                if (results.length >= req.params.curLimit) {
                  res.write('<a href="' + nextURL + '">next</a></td></tr>'); }  
                
                if (!req.params.curDocument) {
                  /**
                   * No selected document
                   * */
                  res.write('</table>');
                  endResponse(res); 
                } else {
                  /**
                   * Selected document
                   * */
                  db.collection(req.params.curCollection, function(err, collection2) {
                    params = {'_id': req.params.curDocument};
                    if (req.params.curDocument.length == 12 || req.params.curDocument.length == 24) { 
                      params = {'_id': new  BSON.ObjectID(req.params.curDocument)}; }
                    collection.find(params).toArray(function(err, results) {
                      path += req.params.curDocument + '/';
                      for (var n in results[0]) { 
                        res.write('' + displayLine(n, results[0][n], path, dbName + '__' + req.params.curCollection + '__' + req.params.curDocument, req.params, true));
                      }
                      res.write('</table>');
                      endResponse(res); 
                    });
                  });
                }
              });
            });
          }
        });
      });
    }
  });
  app.listen(mongoclikkerConnection.web);
  console.log('Listening on port ' + mongoclikkerConnection.web);
};

exports.setHost = funcSetHost;
exports.protectWith = funcProtectWithUser;
exports.andPort = funcAndPort;
exports.withUser = funcWithUser;
exports.andPassword = funcAndPassword;
exports.forDatabase = funcforDatabase;
exports.andListenOn = exports.listenOn = funcAndListenOn;
exports.cast = castData;